### DISCUSSION TITLE
Broken POST routes

### DISCUSSION THREAD

**Discussion by joshhanley** [livewire/livewire#3113] [7. Bugs (archived)]
If you have a page that redirect to a POST route, and that route doesn't return to a GET route (for whatever reason) and you have a Livewire component on it, it throws an error (see screenshot below):

> Symfony\Component\HttpKernel\Exception\MethodNotAllowedHttpException
> The GET method is not supported for this route. Supported methods: POST.

Steps to reproduce

1. Create a normal controller with a get route and a post route that has a completely different URI.
2. Put a form in the get route view that submits to the post route.
2. On the post route create a view that has a Livewire component in it.
3. Try and interact with the Livewire component.

I've created a demo repo here on "post-route-issue" branch
https://github.com/joshhanley/LivewirePlayground/tree/post-route-issue

1. Download repo and composer install.
2. Go to the base route `/` and click on "Submit".
3. Then on the next page, click on "Refresh Server".
4. Error should appear.

Livewire 2.3.15
Laravel 8.26.1

![image](https://user-images.githubusercontent.com/882837/107013480-853a4d00-67ee-11eb-918d-7bb44f0f8c4d.png)

**Comment by labomatik** [comment]
Thanks for @joshhanley for helping reproduce this.

This is what happened after the switch to 2.3.9 and up.
2.3.8 was not triggering the issue.

**Comment by calebporzio** [comment]
Yeah, so in general if you adhere to "PRG" (Post/Redirect/Get) this kind of thing shouldn't happen. (https://en.wikipedia.org/wiki/Post/Redirect/Get)

I do understand that some people don't adhere to the practice.

We have to options as I see it:
A) Adhere to the PRG requirement and document it
B) Perform a check if a route is not found, to scan again with a "POST" version of the route.

I think B is a decent idea. There is already a precedence for this kind of thing with the recent locale re-lookup.

If we do B, do you see any implications? Like could someone use that behavior as a way around the persistent middleware system?

**Comment by joshhanley** [comment]
@calebporzio yeah ideally option A, but as it's only a patch release and it is causing a breaking change for @labomatik I think B is the best option.

I think the implications would be pretty similar to the locale look up, so if there was a way around it, potentially either could cause issues.

**Comment by calebporzio** [comment]
Yep, agreed, thanks @joshhanley

**Comment by labomatik** [comment]
@calebporzio i'm trying to understand what you mean, a POST result should not display anything. is that correct?

**Comment by joshhanley** [comment]
@labomatik yeah Ideally any page that displays information should be on a GET route (like a form). Then you submit the form to the POST route to save stuff.

The POST route, once it has saved, should do a redirect to a GET page (like redirecting back to the original page before the form was submitted, or to another page). But if there were any validation errors, the POST route should redirect back, to the original form, so it can display any errors. That means the POST route will never return any views or actually display anything.

Hope that helps!

**Comment by labomatik** [comment]
perfect, that makes sense, i'll update my code, i was using (wrongly i guess) the post to "hide" the sent data since it's more complicated for a user to replicate a post query than a get

**Comment by calebporzio** [comment]
Ok, this should be fixed in v2.3.17 - thanks!

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by da-sie** [comment]
REOPEN
> B) Perform a check if a route is not found, to scan again with a "POST" version of the route.
> 
> I think B is a decent idea. There is already a precedence for this kind of thing with the recent locale re-lookup.
> 
> If we do B, do you see any implications? Like could someone use that behavior as a way around the persistent middleware system?

Guys, did you just killed by accident my installation?
Since the last update I started to get following errors while trying to open sliding panel with livewire component:
![Zrzut ekranu 2021-02-19 o 15 46 39](https://user-images.githubusercontent.com/459266/108519182-b2482e80-72c9-11eb-8640-7127ba721565.png)

So it works with following route (because there exists GET and POST for /offers):
```
Route::resource('offers', OffersController::class);
```
And it doesn't with this one (GET only):
```
Route::get('offers/show/w/11/1/1/2', [OffersController::class, 'showsample']);
```
If I however change the second to 
```
Route::any('offers/show/w/11/1/1/2', [OffersController::class, 'showsample']);
```
or ugly, but for test purposes:
```
Route::get('offers/show/w/11/1/1/2', [OffersController::class, 'showsample']);
Route::post('offers/show/w/11/1/1/2', [OffersController::class, 'showsample']);
```

Then it works without errors.
Looks like Livewire is checking if the POST exists and if doesn't - it throws an error, but hey, my GET was there and I don't use POST on the url. How are you deciding if GET is not found?!? I have pretty complex route path before the /offers part, with multiple domains and two parameters processed by middlewares, but it worked perfectly two days ago.

**Comment by joshhanley** [comment]
@naszybko what version of Livewire are you on? That GET/POST issue should have been fixed in 2.3.17. Can you let me know?

**Comment by da-sie** [comment]
> @naszybko what version of Livewire are you on? That GET/POST issue should have been fixed in 2.3.17. Can you get me know?

2.3.18 - I will perform tests with previous versions in order to detect last working one.

**Comment by da-sie** [comment]
So it worked perfectly with 2.3.16 and previous versions.
Broke at 2.3.17, followed by broken current - 2.3.18

**Comment by joshhanley** [comment]
@naszybko ok thanks for checking. That's strange I would have expected the opposite, 2.3.16 broken and .17 working. Hmm not sure what might be going on there as it should pick up from the route whether it's a GET or POST.

**Comment by da-sie** [comment]
Ok, I know what happened.
It's all about custom HttpConnectionHandler url...
In our design there are route groups with middlewares that block unauthorized users from accessing certain areas (profiles or shops they don't have access for).

```
use Livewire\Controllers\HttpConnectionHandler as LivewireHandler;

Route::group(['prefix' => 'p,{profile}', 'middleware' => ['profileAccess']], function () {
    Route::group(['prefix' => 's,{shop}', 'middleware' => ['shopAccess'], 'as' => 'shop.'], function () {

        Route::get('sample-works', [SampleController::class, 'index']); //THIS DOES WORK
        Route::post('sample-works', [SampleController::class, 'index']);
        Route::any('sample-works-too', [SampleController::class, 'index']); //THIS DOES WORK TOO
        Route::get('sample', [SampleController::class, 'index']); //THIS DOESN'T WORK
        
        Route::any('livewire/message/{name}', LivewireHandler::class); // MY CUSTOM LIVEWIRE HANDLER ROUTE

    });
});
```

So in these pages we included path to livewire_app_url in order to have HttpConnectionHandler url like this:
p,1/s,1/livewire/message/some.component

```
<script>
window.livewire_app_url = "{{url('p,1/s,1')}}";
</script>
```
...in order to pass livewire requests through these middlewares as additional layer of security and also to let Livewire components access params passed in routes (through middlewares) same way as Laravel controllers do and last, but not least, to work well with mcamara/laravel-localization (the route to livewire_app_url supported language prefixes - reference: https://forum.laravel-livewire.com/t/post-404-with-livewire-mcamara-laravel-localization/1659). 
Yes, it's hacky but worked before introducing v 2.3.17, now it throws these errors.

I've pushed repo with example: https://github.com/naszybko/livewire-get-broken (I only got rid of middlewares from example above) if you want to play with that. Thank you @joshhanley.
